From c1db931022d5fa3f51e865eed4d8c2488bf4af89 Mon Sep 17 00:00:00 2001
From: Valentin Raevsky <valentin@compulab.co.il>
Date: Sun, 29 Mar 2020 13:09:19 +0300
Subject: [PATCH 47/55] cl-som-imx8: ddr: Some small cleanup and refactoring

1) Simplify the lpddr4_desc structure.
2) Clean up the ddr detection logics.

Signed-off-by: Valentin Raevsky <valentin@compulab.co.il>
---
 board/compulab/cl-som-imx8/Makefile  |  5 ++--
 board/compulab/cl-som-imx8/ddr/ddr.c | 48 +++++++++++++++++-------------------
 2 files changed, 25 insertions(+), 28 deletions(-)

diff --git a/board/compulab/cl-som-imx8/Makefile b/board/compulab/cl-som-imx8/Makefile
index 482d77f..5cf1fa1 100644
--- a/board/compulab/cl-som-imx8/Makefile
+++ b/board/compulab/cl-som-imx8/Makefile
@@ -13,13 +13,14 @@ obj-$(CONFIG_POWER_PFUZE100) += ../../freescale/common/pfuze.o
 ifdef CONFIG_SPL_BUILD
 obj-y += spl.o
 ifdef CONFIG_2OP_FULL_SUPPORT
-obj-y += ddr/lpddr4_timing_1g.o ddr/lpddr4_timing_2g.o
 obj-y += ddr/lpddr4_timing_05_10_2g.o
+obj-y += ddr/lpddr4_timing_ff000110_4g_2op.o
 endif
+obj-y += ddr/lpddr4_timing_1g.o
+obj-y += ddr/lpddr4_timing_2g.o
 obj-y += ddr/lpddr4_timing_05_10_2g_3op.o
 obj-y += ddr/lpddr4_timing_1g_3op.o
 obj-y += ddr/lpddr4_timing_2g_3op.o
-obj-y += ddr/lpddr4_timing_ff000110_4g_2op.o
 obj-y += ddr/lpddr4_timing_ff000110_4g_3op.o
 obj-y += ddr/ddr.o
 obj-y += eeprom_spl.o
diff --git a/board/compulab/cl-som-imx8/ddr/ddr.c b/board/compulab/cl-som-imx8/ddr/ddr.c
index d8a2e0b..5a3e6f1 100644
--- a/board/compulab/cl-som-imx8/ddr/ddr.c
+++ b/board/compulab/cl-som-imx8/ddr/ddr.c
@@ -26,16 +26,10 @@ u32 cl_eeprom_get_osize(void);
 unsigned int lpddr4_mr_read(unsigned int mr_rank, unsigned int mr_addr);
 
 struct lpddr4_desc {
-	char name[16];
+	char name[8];
 	unsigned int id;
 	unsigned int size;
-	unsigned int count;
-	/* an optional field
-	 * use it if default is not the
-	 * 1-st array entry */
-	unsigned int _default;
-	struct dram_timing_info *timing[4];
-	char *desc[4];
+	struct dram_timing_info *timing;
 };
 
 struct lpddr4_tcm_desc {
@@ -45,24 +39,27 @@ struct lpddr4_tcm_desc {
 	unsigned int count;
 };
 
-#define DEFAULT (('D' << 24) + ('E' << 16 ) + ( 'F' << 8 ) + 'A')
+#define DEFAULT (('D' << 24) + ('E' << 16) + ('F' << 8) + 'A')
 static const struct lpddr4_desc lpddr4_array_2op[] = {
-	{ .name = "Micron", .id = 0xFF000110, .size = 4096, .count = 1, .timing = { &dram_timing_ff000110_4g_2op } },
+	{ .name = "Samsung",.id = 0x01050008, .size = 1024, .timing = &dram_timing_1g },
+	{ .name = "Nanya",  .id = 0x05000008, .size = 1024, .timing = &dram_timing_1g },
+	{ .name = "Micron", .id = 0xFF020008, .size = 2048, .timing = &dram_timing_2g },
 #if CONFIG_2OP_FULL_SUPPORT
-	{ .name = "Micron", .id = 0xFF020008, .size = 2048, .count = 1, .timing = { &dram_timing_2g } },
-	{ .name = "Samsung",.id = 0x01050008, .size = 1024, .count = 1, .timing = { &dram_timing_1g } },
-	{ .name = "Nanya",  .id = 0x05000008, .size = 1024, .count = 1, .timing = { &dram_timing_1g} },
-	{ .name = "Nanya", . id = 0x05000010, .size = 2048, .count = 1, .timing = { &dram_timing_05_10_2g} },
+	{ .name = "Micron", .id = 0xFF000110, .size = 4096, .timing = &dram_timing_ff000110_4g_2op },
+	{ .name = "Micron", .id = 0xFF020008, .size = 2048, .timing = &dram_timing_2g },
+	{ .name = "Samsung",.id = 0x01050008, .size = 1024, .timing = &dram_timing_1g },
+	{ .name = "Nanya",  .id = 0x05000008, .size = 1024, .timing = &dram_timing_1g },
+	{ .name = "Nanya", . id = 0x05000010, .size = 2048, .timing = &dram_timing_05_10_2g },
 #endif
 };
 
 static const struct lpddr4_desc lpddr4_array_3op[] = {
-	{ .name = "Micron", .id = 0xFF020008, .size = 2048, .count = 1, .timing = { &dram_timing_2g_3op } },
-	{ .name = "Micron", .id = 0xFF000110, .size = 4096, .count = 1, .timing = { &dram_timing_ff000110_4g_3op } },
-	{ .name = "Samsung",.id = 0x01050008, .size = 1024, .count = 1, .timing = { &dram_timing_1g_3op } },
-	{ .name = "Samsung",.id = 0x01061010, .size = 2048, .count = 1, .timing = { &dram_timing_05_10_2g_3op} },
-	{ .name = "Nanya",  .id = 0x05000008, .size = 1024, .count = 1, .timing = { &dram_timing_1g_3op } },
-	{ .name = "Nanya",  .id = 0x05000010, .size = 2048, .count = 1, .timing = { &dram_timing_05_10_2g_3op} },
+	{ .name = "Micron", .id = 0xFF020008, .size = 2048, .timing = &dram_timing_2g_3op },
+	{ .name = "Micron", .id = 0xFF000110, .size = 4096, .timing = &dram_timing_ff000110_4g_3op },
+	{ .name = "Samsung",.id = 0x01050008, .size = 1024, .timing = &dram_timing_1g_3op },
+	{ .name = "Samsung",.id = 0x01061010, .size = 2048, .timing = &dram_timing_05_10_2g_3op },
+	{ .name = "Nanya",  .id = 0x05000008, .size = 1024, .timing = &dram_timing_1g_3op },
+	{ .name = "Nanya",  .id = 0x05000010, .size = 2048, .timing = &dram_timing_05_10_2g_3op },
 };
 
 static unsigned int lpddr4_get_mr(void)
@@ -111,7 +108,7 @@ void spl_dram_init(void)
 	unsigned int ddr_asize = 0;
 	unsigned int ddr_found = 0;
 	int reset_required = 0;
-	int i = 0, j = 0;
+	int i = 0;
 
 	const struct lpddr4_desc *lpddr4_array = NULL;
 	struct lpddr4_tcm_desc *lpddr4_tcm_desc = (struct lpddr4_tcm_desc *) SPL_TCM_DATA;
@@ -153,13 +150,12 @@ void spl_dram_init(void)
 			while ( 1 ) {};
 		}
 		ddr_info = lpddr4_array[i].id;
-	} else
-		j = cl_eeprom_get_drate(0,0) % lpddr4_array[i].count;
+	}
 
-	printf("DDRINFO(%s): %s %dG @ %d MHz\n", (ddr_found ? "D" : "?" ), lpddr4_array[i].name,
-			lpddr4_array[i].size, lpddr4_array[i].timing[j]->fsp_table[0]);
+	printf("DDRINFO(%s): %s %d GiB @ %d MHz\n", (ddr_found ? "D" : "?" ), lpddr4_array[i].name,
+			(lpddr4_array[i].size >> 10), lpddr4_array[i].timing->fsp_table[0]);
 
-	ddr_init(lpddr4_array[i].timing[j]);
+	ddr_init(lpddr4_array[i].timing);
 
 	ddr_info_mrr = lpddr4_get_mr();
 	if (ddr_info_mrr == 0xFFFFFFFF ) {
-- 
1.9.1

